// tide/component.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _shallowequal = require('shallowequal');

var _shallowequal2 = _interopRequireDefault(_shallowequal);

var _lodash = require('lodash.mapvalues');

var _lodash2 = _interopRequireDefault(_lodash);

var _base = require('./base');

var _base2 = _interopRequireDefault(_base);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var NOT_KEY_PATH_PROPS = ['children', 'tide', 'tideOptions', 'key', 'ref'];

function omit(object, func) {
  return Object.keys(object).reduce(function (obj, key) {
    var _extends2;

    var value = object[key];
    return func(key, value) ? obj : _extends({}, obj, (_extends2 = {}, _extends2[key] = value, _extends2));
  }, {});
}

var excludedProps = NOT_KEY_PATH_PROPS.reduce(function (val, prop) {
  val[prop] = true;
  return val;
}, {});

var TideComponent = function (_React$Component) {
  _inherits(TideComponent, _React$Component);

  function TideComponent(props) {
    _classCallCheck(this, TideComponent);

    var _this = _possibleConstructorReturn(this, _React$Component.apply(this, arguments));

    var tide = _this.getTide();
    var keyPaths = _this.getKeyPaths(props, tide);
    _this._componentTide = _extends({
      keyPaths: keyPaths,
      options: _this.getTideOptions()
    }, tide.getComponentProps());

    _this.state = _this.getPropsFromKeyPaths(keyPaths, tide);
    return _this;
  }

  TideComponent.prototype.getChildContext = function getChildContext() {
    return { tide: this.getTide(), tideOptions: this.getTideOptions() };
  };

  TideComponent.prototype.componentWillMount = function componentWillMount() {
    this._onStateChange = this.onStateChange.bind(this);
    this.getTide().onChange(this._onStateChange);
  };

  TideComponent.prototype.componentWillUnmount = function componentWillUnmount() {
    this._isUnmounting = true;
    this.getTide().offChange(this._onStateChange);
  };

  TideComponent.prototype.onStateChange = function onStateChange() {
    if (this._isUnmounting) return;
    var tide = this.getTide();
    var newState = this.getPropsFromKeyPaths(this.getKeyPaths(this.props, tide), tide);
    if (this.hasStaleState(newState)) {
      this.setState(newState);
    }
  };

  TideComponent.prototype.getKeyPaths = function getKeyPaths(props, tide) {
    var _this2 = this;

    var keyPaths = omit(props, function (key) {
      return excludedProps[key];
    });
    keyPaths = (0, _lodash2.default)(keyPaths, function (val, key) {
      var value = typeof val === 'function' ? val(tide.getState(), _this2.getTideOptions()) : val;
      if (Array.isArray(value)) return value;
      if (value === true) return [key];
      return value.split('.');
    });
    return keyPaths;
  };

  TideComponent.prototype.getPropsFromKeyPaths = function getPropsFromKeyPaths(keyPaths, tide) {
    var state = tide.getState();
    return (0, _lodash2.default)(keyPaths, function (value) {
      return state.getIn(value);
    });
  };

  TideComponent.prototype.getTide = function getTide() {
    return this.props.tide instanceof _base2.default ? this.props.tide : this.context.tide;
  };

  TideComponent.prototype.getTideOptions = function getTideOptions() {
    return this.props.tideOptions || this.context.tideOptions || {};
  };

  TideComponent.prototype.getChildProps = function getChildProps() {
    return _extends({}, omit(this.state, function (_, value) {
      return value === undefined;
    }), { tide: this._componentTide });
  };

  TideComponent.prototype.hasStaleState = function hasStaleState(newState) {
    return !(0, _shallowequal2.default)(this.state, newState);
  };

  TideComponent.prototype.render = function render() {
    return this.props.children(this.getChildProps());
  };

  return TideComponent;
}(_react2.default.Component);

exports.default = TideComponent;


if (process.env.NODE_ENV !== 'production') {
  TideComponent.displayName = 'TideComponent';
  TideComponent.propTypes = {
    tide: _propTypes2.default.object
  };
}

var contextTypes = { tide: _propTypes2.default.object, tideOptions: _propTypes2.default.object };
TideComponent.contextTypes = contextTypes;
TideComponent.childContextTypes = contextTypes;
